src/lib.rs

```diff
@@ -83,25 +94,101 @@ fn get_output_mode(args: &Args) -> OutputMode {
     }
 }
 
+struct MaybeInitializedCaptureIndex(AtomicU32);
+
+impl MaybeInitializedCaptureIndex {
+    const UNINITIALIZED: u32 = u32::MAX;
+
+    const FAILED: u32 = u32::MAX - 1;
+
+    fn mark_failed(&self) -> bool {
+        loop {
+            let existing_value = self.0.load(Ordering::Relaxed);
+            if existing_value == Self::FAILED {
+                return false;
+            }
+            let did_store = self.0.compare_exchange(
+                existing_value,
+                Self::FAILED,
+                Ordering::Relaxed,
+                Ordering::Relaxed,
+            );
+            if did_store.is_ok() {
+                return true;
+            }
+        }
+    }
+
+    pub fn get(&self) -> Result<Option<u32>, ()> {
+        let loaded = self.0.load(Ordering::Relaxed);
+        match loaded {
+            loaded if loaded == Self::UNINITIALIZED => Ok(None),
+            loaded if loaded == Self::FAILED => Err(()),
+            loaded => Ok(Some(loaded)),
+        }
+    }
+
+    pub fn get_or_initialize(&self, query: &Query, capture_name: Option<&str>) -> Result<u32, ()> {
+        if let Some(already_initialized) = self.get()? {
+            return Ok(already_initialized);
+        }
+        let capture_index = match capture_name {
+            None => 0,
+            Some(capture_name) => {
+                let capture_index = query.capture_index_for_name(capture_name);
+                if capture_index.is_none() {
+                    let did_mark_failed = self.mark_failed();
+                    if did_mark_failed {
+                        fail(&format!("invalid capture name '{}'", capture_name));
+                    } else {
+                        // whichever other thread "won the race" will have called this fail()
+                        // so we'll be getting killed shortly?
+                        thread::sleep(Duration::from_millis(100_000));
```

Comment:
Let me know if you disagree with this, I am certainly not seeing this cause anything to actually visibly sleep for 100s